lib/article_json/export/common/html/elements/base.rb

```diff
@@ -0,0 +1,85 @@
+module ArticleJSON
+  module Export
+    module Common
+      module HTML
+        module Elements
+          module Base
+            # Extend `base` class with `ClassMethods` upon inclusion
+            def self.included(base)
+              base.extend ClassMethods
+            end
+
+            # @param [ArticleJSON::Elements::Base] element
+            def initialize(element)
+              @element = element
+            end
+
+            # Export a HTML node out of the given element
+            # Dynamically looks up the right export-element-class, instantiates it
+            # and then calls the #build method.
+            # @return [Nokogiri::XML::NodeSet]
+            def export
+              exporter = is_base_class? ? self.class.build(@element) : self
+              exporter.export unless exporter.nil?
+            end
+
+            private
+
+            # Create a Nokogiri NodeSet wrapping a Node with the given `tag`
+            # @param [Symbol] tag - type of the created element
+            # @param [*Object] args - additional arguments for the element
+            # @yield [Nokogiri::XML::Element] Optional block can be passed, will
+            #                                 be executed with the generated
+            #                                 element as a parameter
+            # @return [Nokogiri::XML::NodeSet] Generated main element wrapped in a
+            #                                  NodeSet
+            def create_element(tag, *args)
+              document = Nokogiri::HTML.fragment('').document
+              element = document.create_element(tag.to_s, *args)
+              yield(element) if block_given?
+              wrap_in_node_set([element], document)
+            end
+
+            # Wrap a given list of Nokogiri Nodes in NodeSets
+            # @param [Array[Nokogiri::XML::Node]] elements
+            # @param [Nokogiri::HTML::Document] document
+            # @return [Nokogiri::XML::NodeSet]
+            def wrap_in_node_set(elements, document)
+              Nokogiri::XML::NodeSet.new(document).tap do |node_set|
+                elements.each { |element| node_set.push(element) }
+              end
+            end
+
+            # Check if the current instance is an instance of the base class or
+            # of a specific element. Since this common module is in a different
+            # namespace, a simple `self.class == Base` check does not work.
+            # @return [Boolean]
+            def is_base_class?
+              self.class.name.end_with?('::Base')
+            end
+
+            module ClassMethods
+              # Instantiate the correct sub class for a given element
+              # @param [ArticleJSON::Elements::Base] element
+              # @return [ArticleJSON::Export::Common::HTML::Elements::Base]
```

Comment:
Is this the right way to say "it returns an object implementing this module"?

Reply:
I think so – any other ideas?

Reply:
nope :)

Reply:
for me sounds good